examples/seq2seq/seq2seq.py

```diff
@@ -0,0 +1,301 @@
+            exs = sequence_embed(self.embed_x, xs)
+            h, c, _ = self.encoder(None, None, exs)
+            ys = self.xp.full(batch, EOS, 'i')
+            result = []
+            for i in range(max_length):
+                eys = self.embed_y(ys)
+                eys = F.split_axis(eys, batch, 0)
+                h, c, ys = self.decoder(h, c, eys)
+                cys = F.concat(ys, axis=0)
+                wy = self.W(cys)
+                ys = self.xp.argmax(wy.data, axis=1).astype('i')
+                result.append(ys)
+
+        result = cuda.to_cpu(self.xp.stack(result).T)
+
+        # Remove EOS taggs
+        outs = []
+        for y in result:
+            inds = numpy.argwhere(y == EOS)
+            if len(inds) > 0:
+                y = y[:inds[0, 0]]
+            outs.append(y)
+        return outs
+
+
+def convert(batch, device):
+    def to_device_batch(batch):
+        if device is None:
+            return batch
+        elif device < 0:
+            return [chainer.dataset.to_device(device, x) for x in batch]
+        else:
+            xp = cuda.cupy.get_array_module(*batch)
+            concat = xp.concatenate(batch, axis=0)
+            sections = numpy.cumsum([len(x) for x in batch[:-1]], dtype='i')
+            concat_dev = chainer.dataset.to_device(device, concat)
+            batch_dev = cuda.cupy.split(concat_dev, sections)
+            return batch_dev
+
+    return {'xs': to_device_batch([x for x, _ in batch]),
+            'ys': to_device_batch([y for _, y in batch])}
+
+
+class CalculateBleu(chainer.training.Extension):
+
+    trigger = 1, 'epoch'
+    priority = chainer.training.PRIORITY_WRITER
+
+    def __init__(
+            self, model, test_data, key, batch=100, device=-1, max_length=100):
+        self.model = model
+        self.test_data = test_data
+        self.key = key
+        self.batch = batch
+        self.device = device
+        self.max_length = max_length
+
+    def __call__(self, trainer):
+        with chainer.no_backprop_mode():
+            references = []
+            hypotheses = []
+            for i in range(0, len(self.test_data), self.batch):
+                sources, targets = zip(*self.test_data[i:i + self.batch])
+                references.extend([[t.tolist()] for t in targets])
+
+                sources = [
+                    chainer.dataset.to_device(self.device, x) for x in sources]
+                ys = [y.tolist()
+                      for y in self.model.translate(sources, self.max_length)]
+                hypotheses.extend(ys)
+
+        bleu = bleu_score.corpus_bleu(
+            references, hypotheses,
+            smoothing_function=bleu_score.SmoothingFunction().method1)
+        chainer.report({self.key: bleu})
+
+
+def count_lines(path):
+    with open(path) as f:
+        return sum([1 for _ in f])
+
+
+def load_vocabulary(path):
+    with open(path) as f:
+        return {line.strip(): i for i, line in enumerate(f)}
```

Comment:
Because of `UNK` and `EOS`, `i` -> `i + 2` ?

Reply:
Thank you!